packages/intent-filler/src/strategies/basic.ts

```diff
@@ -0,0 +1,487 @@
+			if (totalNativeTokenNeeded > 0n) {
+				const nativeBalance = await this.provider.getBalance(this.wallet.address)
+
+				// Add some buffer for gas
+				const withGasBuffer = totalNativeTokenNeeded + BigInt(0.001 * 10 ** 18) // 0.001 ETH buffer for gas
+
+				if (BigInt(nativeBalance.toString()) < withGasBuffer) {
+					console.debug(
+						`Insufficient native token balance. Have ${nativeBalance.toString()}, need ${withGasBuffer.toString()}`,
+					)
+					return false
+				}
+			}
+
+			return true
+		} catch (error) {
+			console.error(`Error checking token balances:`, error)
+			return false
+		}
+	}
+
+	/**
+	 * Calculates the total order value for confirmation policy
+	 */
+	private async calculateOrderValue(order: Order): Promise<BigInt> {
+		let totalUSDValue = BigInt(0)
+
+		for (const input of order.inputs) {
+			const tokenUsdPrice = await fetchTokenUsdPriceOnchain(
+				input.token,
+				this.provider,
+				addresses.UniswapV2Router[order.destChain as keyof typeof addresses.UniswapV2Router]!,
+				assets[order.destChain as keyof typeof assets].WETH,
+				assets[order.destChain as keyof typeof assets].USDC,
+			)
+
+			totalUSDValue = totalUSDValue + BigInt(input.amount * BigInt(tokenUsdPrice))
+		}
+
+		return totalUSDValue
+	}
+
+	/**
+	 * Estimates gas for filling an order
+	 */
+	private async estimateGasForFill(order: Order): Promise<BigInt> {
+		try {
+			const contractOrder = this.transformOrderForContract(order)
+
+			const fillOptions: FillOptions = {
+				relayerFee: ethers.utils.parseEther("0.001").toString(),
+			}
+
+			const ethValue = await this.calculateRequiredEthValue(order.outputs)
+
+			const gasEstimate = await this.contract.estimateGas.fillOrder(contractOrder, fillOptions, {
+				value: ethValue,
+			})
+
+			return BigInt(gasEstimate.toString())
+		} catch (error) {
+			console.error(`Error estimating gas:`, error)
+			// Return a conservative estimate if we can't calculate precisely
+			return BigInt(500000)
+		}
+	}
+
+	/**
+	 * Gets the current ETH price in USD
+	 */
+	private async getEthPriceUsd(): Promise<number> {
+		const ethPriceUsd = await fetchTokenUsdPriceOnchain(
+			assets[this.chain as keyof typeof assets].WETH,
+			this.provider,
+			addresses.UniswapV2Router[this.chain as keyof typeof addresses.UniswapV2Router]!,
+			assets[this.chain as keyof typeof assets].WETH,
+			assets[this.chain as keyof typeof assets].USDC,
+		)
+
+		return ethPriceUsd
+	}
+
+	/**
+	 * Gets the HyperBridge protocol fee in ETH
+	 */
+	private async getProtocolFeeEth(order: Order): Promise<number> {
+		const requestBody = this.constructRedeemEscrowRequest(order)
+
+		const dispatchPost: DispatchPost = {
+			dest: order.sourceChain,
+			to: addresses.IntentGateway[order.sourceChain as keyof typeof addresses.IntentGateway]!,
+			body: requestBody,
+			timeout: order.deadline,
+			fee: order.fees,
+			payer: order.user,
+		}
+
+		const protocolFeeEth = await this.contract.quote(dispatchPost)
```

Comment:
this is not the fee in the native token, this is the fee in the feeToken. Configured in the IsmpHost. So you'll have to use uniswapV2 to get the amount in the native token. Alternatively you can just call [`quoteNative`](https://github.com/polytope-labs/ismp-solidity/blob/main/interfaces/IIsmpModule.sol#L136)

See docs for more info: https://docs.hyperbridge.network/developers/evm/fees#native-token-payments